Add Explaination about Parallelize Optimization at FAQ (#3186)
optuna/optuna: docs/source/faq.rst

Soft constraints do not have to be satisfied, but an objective function is penalized if they are unsatisfied. On the other hand, hard constraints must be satisfied.

Optuna is adopting the soft one and **DOES NOT** support the hard one. In other words, Optuna **DOES NOT** have built-in samplers for the hard constraints.

> How can I parallelize optimization

Reviewer: Add `?`.

Author: Thank you for pointing this out. The correction has been addressed.

-------------------------------------------------

The variations that can be expected to be parallelized are in the following three cases.

1. one-node, multi-threading parallelization
2. one-node, multi-processing parallelization

> 3. multi-node, multi-processing parallelization

Reviewer: 1. Multi-threading parallelization with single node
2. Multi-processing parallelization with single node
3. Multi-processing parallelization with multiple nodes

Author: Thank you for pointing this out. The correction has been addressed.

## For the case of 1

> Parallelization can be achieved by setting the argument n_jobs in the optimize method of the Study class.

Reviewer: ```
n_jobs -> `n_jobs`
the Study class -> the :class:`~optuna.study.Study` class
```

Author: Thank you for pointing this out. The correction has been addressed.

> However, since it is multi-threading, the python code will not be faster due to GIL. 

Reviewer: Explain the situation that when we can not fasten the optimization.

Reviewer: c.f. https://en.wikipedia.org/wiki/Global_interpreter_lock

For more information about 1., see APIReference.

> .. APIReference: https://optuna.readthedocs.io/en/stable/reference/index.html

Reviewer: You can check the link is generated by checking the ci of `ci/circleci: document artifact` e.g. 

https://68543-122299416-gh.circle-artifacts.com/0/docs/build/html/index.html

Reviewer: To add links, you need to write those things: https://planset-study-sphinx.readthedocs.io/ja/latest/04.html#id14

```
* Sphinx-users.jp_

.. _Sphinx-users.jp: http://sphinx-users.jp/
```

Author: Thank you for pointing this out. The correction has been addressed.

Soft constraints do not have to be satisfied, but an objective function is penalized if they are unsatisfied. On the other hand, hard constraints must be satisfied.

Optuna is adopting the soft one and **DOES NOT** support the hard one. In other words, Optuna **DOES NOT** have built-in samplers for the hard constraints.

> How can I parallelize optimization?
> -------------------------------------------------

Reviewer 2: Please match the length of "-".
```suggestion
How can I parallelize optimization?
-----------------------------------
```

The variations of parallelization are in the following three cases.

1. Multi-threading parallelization with single node
2. Multi-processing parallelization with single node
3. Multi-processing parallelization with multiple nodes

> 1. Multi-threading parallelization with a single node
> ================================

Reviewer 2: Ditto.
```suggestion
1. Multi-threading parallelization with a single node
=====================================================
```

> =====================================================

Reviewer: ```suggestion
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Reviewer: `=` is already used for the first level indentation. So, we need to use another symbol such as `^`.

Author: Thank you for pointing this out.

> Parallelization can be achieved by setting the argument ``n_jobs`` in :func:`~optuna.study.Study.optimize`.

Author: Thank you.

However, the python code will not be faster due to GIL because :func:`~optuna.study.Study.optimize` with ``n_jobs!=1`` uses multi-threading. 

> While optimizing, it will be faster in limited situations, such as waiting for other server requests, or waiting for C/C++ processing with numpy, etc., but it is not expected to be faster in other cases.

Reviewer: ```suggestion
While optimizing, it will be faster in limited situations, such as waiting for other server requests or C/C++ processing with numpy, etc., but it will not be faster in other cases.
```

For more information about 1., see APIReference_.

> .. _APIReference: https://optuna.readthedocs.io/en/stable/reference/index.html

Reviewer: This link always refer to the stable version. But, each version of the documentation want to refer their version. Can we refer to each version of the page?

Author: >  Can we refer to each version of the page?
Sorry. Where can I find the pages for each version?

For the case of 2

> =================

Reviewer: ```suggestion
^^^^^^^^^^^^^^^^^
```

> For the case of 2
> =================

Reviewer: ```suggestion
2. Multi-processing parallelization with single node
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Reviewer: Ditto.

> For the case of 2
> ================================

Reviewer 2: Ditto.
```suggestion
For the case of 2
=================
```

> This can be achieved by using RDB.

Reviewer: ```suggestion
This can be achieved by using file-based RDBs (e.g. sqlite) and general RDBs (e.g. MySQL).
```

> However, if you cannot install RDB or sqlite, it is not supported, so if you have a request, please make an Issue.

Reviewer 3: I'm afraid that I could not understand this sentence. What do you mean by this sentence?

Author: There seemed to be some inconsistencies in the text, which were pointed out by others, and I have responded to correct them.

> However, if you cannot install RDB or sqlite, it is not supported, so if you have a request, please make an Issue.

Reviewer: ```suggestion
However, if you are in the environment where you can not install an RDB such as sqlite or MySQL, you can not run multi-processing parallelization with single node. When you really want to do it, please request it as a GitHub issue. If we receive a lot of requests, we may provide a solution for it.
```

For more information about 2., see TutorialEasyParallelization_.

.. _TutorialEasyParallelization: https://optuna.readthedocs.io/en/stable/tutorial/10_key_features/004_distributed.html

For the case of 3

> =================

Reviewer: ```suggestion
^^^^^^^^^^^^^^^^^
```

> For the case of 3
> =================

Reviewer: ```suggestion
3. Multi-processing parallelization with multiple nodes
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Reviewer: Ditto.

> For the case of 3
> ================================

Reviewer 2: Ditto.
```suggestion
For the case of 3
=================
```

> This can be achieved by using RDB.
> However, if you cannot install RDB, it is not supported, so if you have a request, please make an Issue.

Reviewer: ```suggestion
This can be achieved by using client/server RDBs (such as PostgreSQL and MySQL).
However, if you are in the environment where you can not install a client/server RDB, you can not run multi-processing parallelization with multiple nodes. When you really want to do it, please request it as a GitHub issue. If we receive a lot of requests, we may provide a solution for it.
```

> However, if you are in the environment where you can not install a client/server RDB, you can not run multi-processing parallelization with multiple nodes. When you really want to do it, please request it as a GitHub issue. If we receive a lot of requests, we may provide a solution for it.

Reviewer 2: The second and third sentences seems to be unnecessary.

```suggestion
However, if you are in the environment where you can not install a client/server RDB, you can not run multi-processing parallelization with multiple nodes.
```

> However, if you are in the environment where you can not install a client/server RDB, you can not run multi-processing parallelization with multiple nodes. When you really want to do it, please request it as a GitHub issue. If we receive a lot of requests, we may provide a solution for it.

Reviewer 2: For multi-process optimization on multiple nodes, it would be impossible in principle not to use RDBs. I think this statement is unnecessary.
